vsintegration/src/FSharp.Editor/Common/LanguageService.fs

```diff
@@ -245,28 +242,38 @@ and
     internal FSharpLanguageService(package : FSharpPackage) =
     inherit AbstractLanguageService<FSharpPackage, FSharpLanguageService>(package)
 
-    let checkerProvider = package.ComponentModel.DefaultExportProvider.GetExport<FSharpCheckerProvider>().Value
     let projectInfoManager = package.ComponentModel.DefaultExportProvider.GetExport<ProjectInfoManager>().Value
 
     let projectDisplayNameOf projectFileName = 
         if String.IsNullOrWhiteSpace projectFileName then projectFileName
         else Path.GetFileNameWithoutExtension projectFileName
 
-    let openedProjects = Queue<IProvideProjectSite>()
+    let singleFileProjects = ConcurrentDictionary<_, AbstractProject>()
 
-    do
-      Events.SolutionEvents.OnAfterOpenProject.Add (fun (args: Events.OpenProjectEventArgs) -> 
-        match args.Hierarchy with
-        | :? IProvideProjectSite as siteProvider ->
-            openedProjects.Enqueue(siteProvider)            
-        | _ -> ())
-        
-    member this.BatchSetupProjects() =    
-        if openedProjects.Count > 0 then    
-            let workspace = package.ComponentModel.GetService<VisualStudioWorkspaceImpl>()
-            for siteProvider in openedProjects do
-                this.SetupProjectFile(siteProvider, workspace)
-            openedProjects.Clear()
+    let tryRemoveSingleFileProject projectId =
+        match singleFileProjects.TryRemove(projectId) with
+        | true, project ->
+            projectInfoManager.RemoveSingleFileProject(projectId)
+            project.Disconnect()       
+        | _ -> ()
+
+    override this.Initialize() =
+        base.Initialize()
+ 
+        this.Workspace.Options <- this.Workspace.Options.WithChangedOption(Completion.CompletionOptions.BlockForCompletionItems, FSharpCommonConstants.FSharpLanguageName, false)
+        this.Workspace.Options <- this.Workspace.Options.WithChangedOption(Shared.Options.ServiceFeatureOnOffOptions.ClosedFileDiagnostic, FSharpCommonConstants.FSharpLanguageName, Nullable false)
+
+        this.Workspace.DocumentClosed.Add <| fun args ->
+            tryRemoveSingleFileProject args.Document.Project.Id
+        //because DocumentClosed doesn't fire when closing solution, we need also this:
+        Events.SolutionEvents.OnAfterCloseSolution.Add <| fun _ ->
+            for id in singleFileProjects.Keys do tryRemoveSingleFileProject id |> ignore
+            
+        Events.SolutionEvents.OnAfterOpenProject.Add <| fun args ->
```

Comment:
Are we adding more work to solution load now? Is there a reason why we're doing this on 'after' open as I assume this runs at "Preparing Solution" time as opposed to the "Loading Project" time.

After the work I put into reducing the time to load F# solutions, it pains me to see us adding more work at load time.

The VisualFSharp is not a representative example of a typical F# solution - at work our F# solution contains >200 projects and takes minutes to load on 12 core, 64GB RAM, SSD machines. If this PR adds 10 seconds to the load time of the VFT solution, then it's probably going to add another 30/40 seconds to larger solutions. Please be extremely wary of this.